ot/unbalanced/_sinkhorn.py

```diff
@@ -865,6 +895,252 @@ def sinkhorn_stabilized_unbalanced(a, b, M, reg, reg_m, reg_type="kl", c=None,
+        c = nx.ones((dim_a, dim_b), type_as=M)
+
+    if n_hists:
+        M0 = M
+    else:
+        c = a[:, None] * b[None, :] if c is None else c
+        M0 = M - reg * nx.log(c)
+    K = nx.exp(-M0 / reg)
+
+    fi_1 = reg_m1 / (reg_m1 + reg) if reg_m1 != float("inf") else 1
+    fi_2 = reg_m2 / (reg_m2 + reg) if reg_m2 != float("inf") else 1
+
+    k1 = reg * reg_m1 / ((reg + reg_m1) * (reg_m1 + reg_m2)) if reg_m1 != float("inf") else 0
+    k2 = reg * reg_m2 / ((reg + reg_m2) * (reg_m1 + reg_m2)) if reg_m2 != float("inf") else 0
+
+    k_rho1 = k1 * reg_m1 / reg if reg_m1 != float("inf") else 0
+    k_rho2 = k2 * reg_m2 / reg if reg_m2 != float("inf") else 0
+
+    if reg_m1 == float("inf") and reg_m2 == float("inf"):
+        xi1, xi2 = 0, 0
+        fi_12 = 1
+    elif reg_m1 == float("inf"):
+        xi1 = 0
+        xi2 = reg / reg_m2
+        fi_12 = reg_m2
+    elif reg_m2 == float("inf"):
+        xi1 = reg / reg_m1
+        xi2 = 0
+        fi_12 = reg_m1
+    else:
+        xi1 = (reg_m2 * reg) / (reg_m1 * (reg + reg_m1 + reg_m2))
+        xi2 = (reg_m1 * reg) / (reg_m2 * (reg + reg_m1 + reg_m2))
+        fi_12 = reg_m1 * reg_m2 / (reg_m1 + reg_m2)
+
+    xi_rho1 = xi1 * reg_m1 / reg if reg_m1 != float("inf") else 0
+    xi_rho2 = xi2 * reg_m2 / reg if reg_m2 != float("inf") else 0
+
+    reg_ratio1 = reg / reg_m1 if reg_m1 != float("inf") else 0
+    reg_ratio2 = reg / reg_m2 if reg_m2 != float("inf") else 0
+
+    err = 1.
+
+    for i in range(numItermax):
+        uprev = u
+        vprev = v
+
+        Kv = nx.dot(K, v_)
+        u_hat = (a / Kv) ** fi_1 * nx.sum(b * v_**reg_ratio2)**k_rho2
+        u_ = u_hat * nx.sum(a * u_hat**(-reg_ratio1))**(-xi_rho1)
+
+        Ktu = nx.dot(K.T, u_)
+        v_hat = (b / Ktu) ** fi_2 * nx.sum(a * u_**(-reg_ratio1))**k_rho1
+        v_ = v_hat * nx.sum(b * v_hat**(-reg_ratio2))**(-xi_rho2)
+
+        t = (nx.sum(a * u_**(-reg_ratio1)) / nx.sum(b * v_**(-reg_ratio2)))**(fi_12 / reg)
+
+        u = u_ * t
+        v = v_ / t
+
+        err_u = nx.max(nx.abs(u - uprev)) / max(
+            nx.max(nx.abs(u)), nx.max(nx.abs(uprev)), 1.
+        )
+        err_v = nx.max(nx.abs(v - vprev)) / max(
+            nx.max(nx.abs(v)), nx.max(nx.abs(vprev)), 1.
+        )
+        err = 0.5 * (err_u + err_v)
+        if log:
+            dict_log['err'].append(err)
+            if verbose:
+                if i % 50 == 0:
+                    print(
+                        '{:5s}|{:12s}'.format('It.', 'Err') + '\n' + '-' * 19)
+                print('{:5d}|{:8e}|'.format(i, err))
+
+        if err < stopThr:
+            break
+
+    if log:
+        dict_log['logu'] = nx.log(u + 1e-300)
+        dict_log['logv'] = nx.log(v + 1e-300)
+
+    if n_hists:  # return only loss
+        res = nx.einsum('ik,ij,jk,ij->k', u, K, v, M)
+        if log:
+            return res, dict_log
+        else:
+            return res
```

Comment:
this line was not tested

Reply:
I called the wrong function in `sinkhorn_unbalanced2`. It should be ok now.

Reply:
This is now tested in the function `test_unbalanced_multiple_inputs`, but it seems there is a problem with `sinkhorn_unbalanced2` when there is only one histogram b. It does not pass the test for any of the methods.

Reply:
It is ok now. The problem was that the reg_type "kl" is not implemented yet for multiple inputs.